Std/Tactic/Change.lean

```diff
@@ -0,0 +1,111 @@
+For example, having `k` return `fvars.map .some` causes all reverted variables to be
+`intro`ed and linked.
+
+Returns the value returned by `k` along with the resulting goal.
+ -/
+def withReverted (mvarId : MVarId) (fvarIds : Array FVarId)
+    (k : MVarId → Array FVarId → MetaM (α × Array (Option FVarId) × MVarId))
+    (clearAuxDeclsInsteadOfRevert := false) : MetaM (α × MVarId) := do
+  let (xs, mvarId) ← mvarId.revert fvarIds true clearAuxDeclsInsteadOfRevert
+  let (r, xs', mvarId) ← k mvarId xs
+  let (ys, mvarId) ← mvarId.introNP xs'.size
+  mvarId.withContext do
+    for x? in xs', y in ys do
+      if let some x := x? then
+        pushInfoLeaf (.ofFVarAliasInfo { id := y, baseId := x, userName := ← y.getUserName })
+  return (r, mvarId)
+
+/--
+Replace the type of the free variable `fvarId` with `typeNew`.
+
+If `checkDefEq = true` then throws an error if `typeNew` is not definitionally
+equal to the type of `fvarId`. Otherwise this function assumes `typeNew` and the type
+of `fvarId` are definitionally equal.
+
+This function is the same as `Lean.MVarId.changeLocalDecl` but makes sure to push substitution
+information into the infotree.
+-/
+def changeLocalDecl' (mvarId : MVarId) (fvarId : FVarId) (typeNew : Expr)
+    (checkDefEq := true) : MetaM MVarId := do
+  mvarId.checkNotAssigned `changeLocalDecl
+  let (_, mvarId) ← mvarId.withReverted #[fvarId] fun mvarId fvars => mvarId.withContext do
+    let check (typeOld : Expr) : MetaM Unit := do
+      if checkDefEq then
+        unless ← isDefEq typeNew typeOld do
+          throwTacticEx `changeLocalDecl mvarId
+            m!"given type{indentExpr typeNew}\nis not definitionally equal to{indentExpr typeOld}"
+    let finalize (targetNew : Expr) := do
+      return ((), fvars.map .some, ← mvarId.replaceTargetDefEq targetNew)
+    match ← mvarId.getType with
+    | .forallE n d b bi => do check d; finalize (.forallE n typeNew b bi)
+    | .letE n t v b ndep => do check t; finalize (.letE n typeNew v b ndep)
+    | _ => throwTacticEx `changeLocalDecl mvarId "unexpected auxiliary target"
+  return mvarId
+
+end Lean.MVarId
+
+open Lean Elab Tactic Meta
+
+/-- `change` can be used to replace the main goal or its local
+variables with definitionally equal ones.
+
+For example, if `n : ℕ` and the current goal is `⊢ n + 2 = 2`, then
+```lean
+change _ + 1 = _
+```
+changes the goal to `⊢ n + 1 + 1 = 2`. The tactic also applies to the local context.
+If `h : n + 2 = 2` and `h' : n + 3 = 4` are in the local context, then
+```lean
+change _ + 1 = _ at h h'
+```
+changes their types to be `h : n + 1 + 1 = 2` and `h' : n + 2 + 1 = 4`.
+
+Change is like `refine` in that every placeholder needs to be solved for by unification,
+but you can use named placeholders and `?_` where you want `change` to create new goals.
+
+The tactic `show e` is interchangeable with `change e`, where the pattern `e` is applied to
+the main goal. -/
```

Comment:
Would it be better to enhance `show` and migrate code using `change` to it?  I saw the comment how it was using `show` behind the scenes in a hacky way.

Reply:
I like the idea of not having two synonymous tactics, but `show T at h` doesn't make a whole lot of sense linguistically.

Reply:
Linguistically, one option would be `change to T` and `change h to T`.  It's a little bit more typing so the `to` could be optional if `h` is omitted.

Reply:
The issue there is that there is *also* a clear pattern in lean tactics of using `at` for referring to hypotheses, so if we were to change the syntax I would rather we do something more syntactic like `change h : T` or similar.

Reply:
I have no objection to the `change h : T` (or `show h : T`) syntax.  I also am ok with keeping the syntax `change T at h`.

My main concern is I'd like to not unintentional provide two ways of writing the same thing so that users aren't confused about which one to use.

Reply:
Well, all of these tactics exist as logical completions of other things, so we end up with two tactics that mean the same thing (and say as much in the documentation). It might be good to bring this up on zulip.

Reply:
It seems like the options here are:
* leave as is, with overlap between the tactics
* remove `change` entirely (i.e. remove `change` from Lean, and `change at` from Mathlib/Std)

I would be happy with either, but I'd prefer to do it separately with the PR that moves things around.